### DISCUSSION TITLE
How to get the permissions of a pdf document

### DISCUSSION THREAD

**Discussion by vrrashkov** [py-pdf/pypdf#2389] [Q&A] [answered]
Probably I am missing it, but I cannot find how to retrieve the permissions of a PDF document, the only option that I see in the documentation is `decode_permissions` but this only returns the permission from the argument that I provide it with. 

Is there something that will return me all the permissions or a way to get the "permissions_code" so I can pass it to the decode_permissions of the current document that I have loaded?

**Comment by stefan6419846** [comment, marked as answer]
Update: Starting at version 4.0.0, you can now use `reader.user_access_permissions` directly.

---

### Old answer

Skimming through the code, there has never been a clean API for this.

`reader._encryption.values.Perms` holds the permission handlers (PDF 1.7, section 8.7.3).

According to the PDF docs, the `P` key of the encryption dictionary holds the corresponding flags. You can access the flags dictionary by this with `reader.decode_permissions(reader._encryption.P)`. Please note that this is an internal attribute, thus this code might break in the future.

**Reply by vrrashkov** [reply]
Thank you for the descriptive answer, It's a bit strange that this is not cleanly provided from the API. The answer you have me is exactly what I need, even if it changes in the future it won't be a problem to update on my end.

**Reply by stefan6419846** [reply]
As a heads-up: For pypdf 4.0.0, `decode_permissions` has been deprecated in favor of `user_access_permissions` to provide a clean API. To get some similar output as before, you probably want to use `reader.user_access_permissions.to_dict()` to get a dictionary-based representation of the corresponding flags.
